Add docs for GraphBLAS dialect ops

paul-tqh-nguyen opened the pull request:
This PR adds basic documentation explaining how the ops in the GraphBLAS dialect work. 

NB: These changes require the changes in https://github.com/metagraph-dev/pymlir/pull/16.
NB: This PR includes a minor tweak to `mlir_graphblas/src/lib/GraphBLAS/GraphBLASOps.cpp`.
NB: This PR includes minor documentation changes to `mlir_graphblas/src/include/GraphBLAS/GraphBLASOps.td`.

There are a few commits here, so let's squash them if this PR is merge-worthy.

jim22k commented on docs/dialect/graphblas_dialect_tutorials/graphblas_lower/graphblas_matrix_reduce_to_scalar.ipynb:
```diff
@@ -0,0 +1,341 @@
+{
+ "cells": [
```
I feel like this whole notebook could be a removed and a short explanation of why this fused operation exists -- to drop a lot of unused code. Going through an example to show that the function works as expected doesn't add any additional information. Telling them it is equivalent (but more efficient) than `matrix_multiply` followed by `matrix_reduce_to_scalar` is sufficient.

You can also tell them that normally users don't call this directly, but the optimizer will find and convert candidate sequences of graphblas calls automatically.

paul-tqh-nguyen replied:
https://github.com/metagraph-dev/mlir-graphblas/pull/85/commits/9799d7fb69b4a15d050b93168a2ac61ba1ea06e3 implements this suggestion. Let me know if this second draft is any better.

jim22k commented on docs/dialect/graphblas_dialect_tutorials/graphblas_optimize/fuse_select.ipynb:
```diff
@@ -0,0 +1,340 @@
+   "id": "450ca5d1",
+   "metadata": {},
+   "source": [
+    "This is difficult to read due to the verbosity of the sparse tensor encodings. Let's make it more terse. "
```
You already explained above, plus the name is self-explanatory. Just use the `tersify_mlir` function the first time when looking at the output result.

paul-tqh-nguyen replied:
https://github.com/metagraph-dev/mlir-graphblas/pull/85/commits/420e21b4cff27114fab3bd8eff9bbfe992134fed implements this suggestion.

jim22k commented on docs/dialect/graphblas_dialect_tutorials/graphblas_optimize/fuse_select.ipynb:
```diff
@@ -0,0 +1,340 @@
+   "id": "3510cff0",
+   "metadata": {},
+   "source": [
+    "## Fusing Single Selector or Multi-Selector graphblas.matrix_select Ops\n",
```
This 2nd section may not be needed. Most users would expect this to work same as above.

The 3rd section is useful in that it shows how fusion is intelligent in inspecting not just sequences of calls, but the inputs as well to make sure the end result yields a correct answer.

paul-tqh-nguyen replied:
I like being more explicit in showing that the fusion also works in the case where we have both single selector and multi-selector ops, so I chose to get rid of the first section and make the second section the primary example. I don't think it's much more complex and gets the idea across. 

I implemented this in https://github.com/metagraph-dev/mlir-graphblas/pull/85/commits/b8074dcc2de4c2509dec1c78e27e048839cba80c.

Let me know what you think.

jim22k reviewed (commented):
Overall looks good. A bit verbose. Maybe see if some things could be trimmed.

For example, showing the non-mask and mask version of matrix multiply is useful a few times, but is not needed every time.

Showing the results are correct at the end of each op section is not needed. It feels more like something you would do in unit testing rather than in documentation. The goal is to show usage, not correctness of implementation to the user.

paul-tqh-nguyen commented on docs/dialect/ops_reference.rst:
```diff
@@ -0,0 +1,478 @@
+given, in which case multiple results will be returned The given sparse tensor
+must be a matrix, i.e. have rank 2. The input tensor must have a CSR sparsity
+or a CSC sparsity. The resulting sparse tensors will have the same sparsity as
+the given sparse tensor.
```
Rather than updating the documentation in several places only to have to revert it back later, I'm going to leave this inaccuracy here and leave the issue of addressing the incorrect behavior of "triu" and "tril" to be handled in https://github.com/metagraph-dev/mlir-graphblas/issues/88.

paul-tqh-nguyen commented on docs/dialect/ops_reference.rst:
```diff
@@ -0,0 +1,478 @@
+.. code-block:: text
+
+    %thunk = constant 100 : i64
+    %answer = graphblas.matrix_apply %sparse_tensor, %thunk { apply_operator = "min" } : (tensor<2x3xi64, #CSR64>, i64) to tensor<2x3xi64, #CSR64>
```
https://github.com/metagraph-dev/mlir-graphblas/pull/85/commits/148757fc7b3b5729c4f201dee8258da1edf26a93 makes the sphinx documentation and the documentation in `mlir_graphblas/src/include/GraphBLAS/GraphBLASOps.td` use variable sized tensors.

paul-tqh-nguyen commented on docs/dialect/graphblas_dialect_tutorials/graphblas_lower/graphblas_convert_layout.ipynb:
```diff
@@ -0,0 +1,628 @@
+   "source": [
+    "The GraphBLAS dialect's ops operate on and usually return sparse matrices. To analyze the results of the ops, it'd be nice to have some functions that'll convert those sparse matrices into dense matrices. The MLIR code below does that. \n",
+    "\n",
+    "We'll have 2 functions, `csr_densify4x4` and `csc_densify4x4`, which will return dense matrices from 4x4 sparse matrices in CSR and CSC formats, respectively. "
```
Since this'll require some big changes, I've moved this to https://github.com/metagraph-dev/mlir-graphblas/issues/89 so that it can be handled in another PR.

paul-tqh-nguyen commented on docs/dialect/graphblas_dialect_tutorials/graphblas_lower/graphblas_matrix_select.ipynb:
```diff
@@ -0,0 +1,847 @@
+    "- Each result matrix has the same size as the given matrix.\n",
+    "- Each result matrix contains only a subset of the elements in the given sparse matrix\n",
+    "- Which elements are kept in each result matrix is determined by a given `selectors` attribute. \n",
+    "\n",
```
Yes, it turns out I needed a newline before the bulleted list. I've implemented this suggestion in https://github.com/metagraph-dev/mlir-graphblas/pull/85/commits/3a990078856fe8d39283c72240cd2714c00cbde2.

paul-tqh-nguyen commented on docs/dialect/graphblas_dialect_tutorials/graphblas_lower/graphblas_matrix_select.ipynb:
````diff
@@ -0,0 +1,847 @@
+    "Here's an example use of the `graphblas.matrix_select` op:\n",
+    "```\n",
+    "%answers = graphblas.matrix_select %sparse_tensor { selectors = [\"triu\", \"tril\"] } : tensor<?x?xf64, #CSR64> to tensor<?x?xf64, #CSR64>\n",
+    "```\n",
````
I've implemented this suggestion in https://github.com/metagraph-dev/mlir-graphblas/pull/85/commits/69228147b11013b344bd3c8bc8cc6c27fd496cbb.

paul-tqh-nguyen commented on docs/dialect/graphblas_dialect_tutorials/graphblas_lower/graphblas_matrix_select.ipynb:
```diff
@@ -0,0 +1,847 @@
+    "- `triu`: Select the elements in the upper triangle.\n",
+    "- `tril`: Select the elements in the lower triangle.\n",
+    "- `gt0`: Select the elements with values greater than 0.\n",
+    "\n",
```
Yes, it was a white space issue with markdown. https://github.com/metagraph-dev/mlir-graphblas/commit/3a990078856fe8d39283c72240cd2714c00cbde2 implements this suggestion.

paul-tqh-nguyen commented on docs/dialect/graphblas_dialect_tutorials/graphblas_lower/graphblas_matrix_multiply.ipynb:
```diff
@@ -0,0 +1,873 @@
+    "sizes = np.array([4, 4], dtype=np.uint64)\n",
+    "sparsity = np.array([False, True], dtype=np.bool8)\n",
+    "\n",
+    "mask = mlir_graphblas.sparse_utils.MLIRSparseTensor(indices, values, sizes, sparsity)"
```
That's a great idea. https://github.com/metagraph-dev/mlir-graphblas/pull/85/commits/ba76211656fa5079681aa44aaa78908234bbdd9f implements this suggestion. I kept the dtype as `np.float64` since we don't support boolean sparse tensors yet.

paul-tqh-nguyen commented:
@jim22k, I implemented your suggestions where possible, moved some suggestions out to separate issues, and made some doc changes whose improvement level may be subjective. I think it's ready for a second review.
